File: lib-database/src/main/java/com/ustadmobile/core/db/dao/ConnectivityStatusDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmInsert;
import com.ustadmobile.lib.database.annotation.UmOnConflictStrategy;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmTransaction;
import com.ustadmobile.lib.db.entities.ConnectivityStatus;

@UmDao
public abstract class ConnectivityStatusDao{

    @UmInsert (onConflict = UmOnConflictStrategy.REPLACE)
    public abstract void insert(ConnectivityStatus connectivityStatus, UmCallback<Long> callback);

    @UmInsert(onConflict =  UmOnConflictStrategy.REPLACE)
    public abstract void insert(ConnectivityStatus connectivityStatus);

    @UmQuery("UPDATE ConnectivityStatus SET connectivityState = :connectivityState")
    public abstract void updateState(int connectivityState, UmCallback<Void> callback);

    @UmQuery("UPDATE ConnectivityStatus SET connectivityState = :connectivityState , wifiSsid = :wifiSsid")
    public abstract void updateState(int connectivityState, String wifiSsid, UmCallback<Void> callback);

    @UmQuery("UPDATE ConnectivityStatus SET connectivityState = :connectivityState , wifiSsid = :wifiSsid")
    public abstract void updateStateSync(int connectivityState, String wifiSsid);

    public void addConnectivityStatusRecord(int state, String wifiSsid, boolean connectedOrConnecting,
                                            UmCallback<Void> callback){
        ConnectivityStatus connectivityStatus = new ConnectivityStatus();
        connectivityStatus.setConnectedOrConnecting(connectedOrConnecting);
        connectivityStatus.setConnectivityState(state);
        connectivityStatus.setWifiSsid(wifiSsid);
        insert(connectivityStatus,null);
    }

    @UmQuery("SELECT ConnectivityStatus.* FROM ConnectivityStatus LIMIT 1")
    public abstract UmLiveData<ConnectivityStatus> getStatusLive();

    @UmQuery("SELECT ConnectivityStatus.* FROM ConnectivityStatus LIMIT 1")
    public abstract ConnectivityStatus getStatus();
}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/ConnectivityStatusDao.kt
package com.ustadmobile.core.db.dao

import androidx.room.*
import com.ustadmobile.door.annotation.DoorDao
import kotlinx.coroutines.flow.Flow
import com.ustadmobile.lib.db.entities.ConnectivityStatus

@DoorDao
expect abstract class ConnectivityStatusDao {

    @Query("SELECT ConnectivityStatus.* FROM ConnectivityStatus LIMIT 1")
    abstract fun statusLive(): Flow<ConnectivityStatus?>

    @Query("SELECT ConnectivityStatus.* FROM ConnectivityStatus LIMIT 1")
    abstract fun status(): ConnectivityStatus?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insert(connectivityStatus: ConnectivityStatus): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract suspend fun insertAsync(connectivityStatus: ConnectivityStatus): Long

    @Query("UPDATE ConnectivityStatus SET connectivityState = :connectivityState")
    abstract suspend fun updateStateAsync(connectivityState: Int)

    @Query("UPDATE ConnectivityStatus SET connectivityState = :connectivityState , wifiSsid = :wifiSsid")
    abstract suspend fun updateState(connectivityState: Int, wifiSsid: String)

    @Query("UPDATE ConnectivityStatus SET connectivityState = :connectivityState , wifiSsid = :wifiSsid")
    abstract fun updateStateSync(connectivityState: Int, wifiSsid: String)

}
